### ISSUE TITLE
New DOM renderer for Flutter web

### ISSUE THREAD

**Issue by mgazisalik** [flutter/flutter#139292]
# Introduction
It would be great a DOM renderer into the Flutter framework for building websites. The primary goal is to expand Flutter's web capabilities by improving SEO, enabling server-side rendering, and ensuring compatibility with popular front-end tools such as Vite.

### Related Issues
Issue #46789: Improve SEO for Flutter web apps
Issue #47600: Add support for server-side rendering in Flutter for web

## Proposed Solution
The main idea is to introduce a new **'runPage'** method that directly renders components to the DOM instead of Flutter's existing engine. Key aspects:

- Components designed specifically for DOM rendering
- Support for Client-Side Rendering (CSR), Server-Side Rendering (SSR), Incremental Static Regeneration (ISR)
- Hybrid rendering design for routes
- Widgets and dom elements interoperability support
- Compatibility with popular front-end tools

**Comment by MarkOSullivan94** [comment]
For people who are wanting this have you considered [Jaspr](https://github.com/schultek/jaspr) as an alternative?

**Comment by jsloop42** [comment]
> For people who are wanting this have you considered [Jaspr](https://github.com/schultek/jaspr) as an alternative?

In my use case it's not a standalone web app. We are developing for iOS, Android and Web. We use the same code base and sometimes different widget specific for web layout.

**Comment by miyoyo** [comment]
My fundamental disagreement with this proposal is more of a question of scope than a question of possibility.

There seems to be two part to this request:
1. Editing the surrounding HTML that hosts the canvas flutter draws on
  * No issue with that, really
2. *Using flutter to render DOM components*, or a hybrid dom/canvas implementation
  * Now that's where the conflict is

Is it possible to make a full flutter implementation in pure dom that can be serialized and sent to the client for instant render? **Maybe.**

Is it possible to make a fully hybrid dom/canvas intermingle that also respects CSS, Z Index and two not-exactly-the-same constraint and sizing systems? **Maybe.**

Is it something flutter should do? **I don't believe so.**

This is not just a small request here, this is basically asking the flutter team to reinvent web *once again*, but this time to integrate and work around the limitations and the requirements of a full blown web document, the initial idea behind flutter being *not using the damn thing in the first place.*

Flutter is great, don't get me wrong, I love it, but trying to be "the everyframework" in every single possible way will either result in it stalling because of no focus, or ending up with a terrible compromise on all platforms.

**Comment by mrgzi** [comment]
> Flutter is great, don't get me wrong, I love it, but trying to be "the everyframework" in every single possible way will either result in it stalling because of no focus, or ending up with a terrible compromise on all platforms.

I disagree with this. Flutter already supports many platforms, even for embedded devices. So far this didn't make the flutter "the everyframework" but the target website will make it? Websites are the last thing Flutter needs to start supporting

**Comment by oravecz** [comment]
> > Flutter is great, don't get me wrong, I love it, but trying to be "the everyframework" in every single possible way will either result in it stalling because of no focus, or ending up with a terrible compromise on all platforms.
> 
> I disagree with this. Flutter already supports many platforms, even for embedded devices. So far this didn't make the flutter "the everyframework" but the target website will make it? Websites are the last thing Flutter needs to start supporting

I think you drastically underestimate the amount of resources necessary to accomplish this. I am hoping that the new WASM model is highly prioritized and no resources are spent chasing this I’ll-fated concept.

To the people who want the Flutter SDK to generate a pure DOM output, put together a POC to show how the render objects can output this structure. I suspect there is a really good reason why the existing DOM renderer produces the output it does.

**Comment by miyoyo** [comment]
> > Flutter is great, don't get me wrong, I love it, but trying to be "the everyframework" in every single possible way will either result in it stalling because of no focus, or ending up with a terrible compromise on all platforms.
> 
> I disagree with this. Flutter already supports many platforms, even for embedded devices. [...]  Websites are the last thing Flutter needs to start supporting

Because the proposal doesn't ask to just support another platform, Flutter already runs on the web.

Flutter is *intentionally* a framework that brings everything along (it's own language, runtime, engine, renderer, etc.) and only providing a shim to the platform to ensure that the same code will run in the same way doing the same thing on all platforms.

[As you specified on your reddit post](https://www.reddit.com/r/FlutterDev/comments/187mdit/comment/kbg2yyy/?utm_source=share&utm_medium=web2x&context=3), this proposal is not asking to merely add a part to flutter, but to add a completely different way of working.
```dart
void main() {
  if (kIsWeb) {
    runPage(MainPage(child: Body()));
  } else {
    runApp(MainApp(child: MaterialApp()));
  }
}
```

This completely goes against flutter's vision, it's asking to add **web component support** in a brand new, separate part of the code specifically just to deal with web, it's asking flutter to just be another React *on top* of what it already is, and, of course, none of these elements will be usable on other platforms *than* web.

<details><summary>Of course, we can go full reductio ad absurdum</summary>
<p>
Why not add support for Android's native UI? And iOS? And Windows? And macOS? And Linux?

```dart
void main() {
  if (kIsWeb) {
    runPage(MainPage(child: Body()));
  } else if (isAndroid) {
    runAndroidActivity(child: Activity());
  } else if (isIOS) {
    runViewController(...);
  } else if (isWindows) {
    runWinUIWindow(...);
  } else if (...) {
    // you catch my drift
  } else {
    runApp(MainApp(child: MaterialApp()));
  }
}
```

[I wonder if there's another tool whose entire point is literally that.](https://reactnative.dev/)

</p>
</details>

**Comment by mrgzi** [comment]
> Because the proposal doesn't ask to just support another platform, Flutter already runs on the web.

Yes, but we don't take advantage of the full web capability. Just because Flutter works as an application on the web does not mean that its completely dominates the web. This is like being able to render widgets to Android or iOS but not being able to use native features. In the same way, Flutter web limits us and we cannot benefit from the features of the web.

> Why not add support for Android's native UI? And iOS? And Windows? And macOS? And Linux?

Because we don't need native UI thanks to flutter engine. But we need dom elements because it is impossible for flutter engine to provide this.

**Comment by miyoyo** [comment]
> Yes, but we don't take advantage of the full web capability.

Just like we don't take full advantage of Windows' capabilities, or macOS', or iOS, or Android, or any platform really.

> Just because Flutter works as an application on the web does not mean that its completely dominates the web

Who said it should dominate the web? It's a tool, tools don't have to do everything.

> This is like being able to render widgets to Android or iOS but not being able to use native features.

No, this could have been an argument if you talked about platform views [before they were implemented on the web](https://docs.flutter.dev/release/breaking-changes/platform-views-using-html-slots-web), but this is not valid anymore.

> Because we don't need native UI thanks to flutter engine. But we need dom elements because it is impossible for flutter engine to provide this.

I could use the exact same argument in reverse if this proposal was about supporting SwiftUI:
"Why not add support for the Web?"
"Because we don't need web UI thanks to flutter engine. But we need SwiftUI elements because it is impossible for flutter engine to provide this."

To reiterate, I'm not against having some additional support of the surrounding HTML to support things like SEO or metadata.
I do think that it's not justifiable to ask for a whole new UI system alongside widgets to specifically provide alternative ways to do UI for one platform, alternative ways that will never have feature parity on all platforms.

**Comment by jodinathan** [comment]
We need a front framework on top of Flutter that prints Widgets or DOM elements when necessary.  
A bright new framework that won't work with currently Flutter apps and with very limited initial components. Then, if it has traction, we would get more components.

That can be done when Static meta programming is available.

**Comment by mrgzi** [comment]
> Just like we don't take full advantage of Windows' capabilities, or macOS', or iOS, or Android, or any platform really.

Yes but at least developers can write their libraries to extend those platform's capabilities. In the current situation for flutter web, we can't do it for websites.

**Comment by slavap** [comment]
@yjbanov It would be much better if some resources were assigned to just fixing bugs in the existing HTML renderer - it sounds like giving false hope to the Flutter Web developers. 
Also already mentioned https://github.com/schultek/jaspr may be the way to go for a "regular" DOM web app - though code automatically becomes non-cross-platform.

**Comment by miyoyo** [comment]
> > Just like we don't take full advantage of Windows' capabilities, or macOS', or iOS, or Android, or any platform really.
> 
> Yes but at least developers can write their libraries to extend those platform's capabilities. In the current situation for flutter web, we can't do it for websites.

You can do just as much with platform channels and platform views in Android, iOS, etc, than you can do with the web version.
If you don't agree with this, please give an exact example of something that's wholly impossible on the web, but doable on other platforms.

The only thing that could be justified is allowing people to override parts of the HTML without flutter overwriting it, but that falls into the SEO/Metadata discussion.

**Comment by mrgzi** [comment]
> You can do just as much with platform channels and platform views in Android, iOS, etc, than you can do with the web version.
If you don't agree with this, please give an exact example of something that's wholly impossible on the web, but doable on other platforms, and that isn't an inherent limitation of browsers (like TCP/UDP Communication for example)

I meant that you can even write low-level code for those platforms and interop with dart language. So this technically means that a lot of taking benefit of platforms. But in flutter web, we don't take benefit as much as like those platforms.

**Comment by miyoyo** [comment]
> > You can do just as much with platform channels and platform views in Android, iOS, etc, than you can do with the web version.
> > If you don't agree with this, please give an exact example of something that's wholly impossible on the web, but doable on other platforms, and that isn't an inherent limitation of browsers (like TCP/UDP Communication for example)
> 
> I meant that you can even write low-level code for those platforms and interop with dart language. So this technically means that a lot of taking benefit of platforms. But in flutter web, we don't take benefit as much as like those platforms.

That isn't an example, you can write javascript in packages and use it in your code via platform channels.
You can use HTML elements via platform channels.
You can ship JS alongside your compiled flutter web app, hell, you can technically edit the html file it builds too.

Be specific, give an example of a case where this is not a problem in phones and desktops, but it is on the web, give code examples, or refer to exact issues, but don't just handwave like that.

**Comment by mrgzi** [comment]
I love flutter and writing a UI is very easy and fun. I would like to get same benefit for websites. For example; to write a large website such as reddit, with flutter engine will not be useful. The main thing is here not just seo, engine initialization is slow.

**Comment by mrgzi** [comment]
@miyoyo I think this feature is not like pushing every feature to a programming language and over time make the language ugly. With a Dom renderer, flutter will be fully cross-platform.

**Comment by miyoyo** [comment]
> I love flutter and writing a UI is very easy and fun. I would like to get same benefit for websites. For example; to write a large website such as reddit, with flutter engine will not be useful. The main thing is here not just seo, engine initialization is slow.
> 
> @miyoyo I think this feature is not like pushing every feature to a programming language and over time make the language ugly. With a Dom renderer, flutter will be fully cross-platform.

So you're not asking for flutter, just something that feels like flutter.

Just like people have been saying twice already, [Try Jaspr](https://github.com/schultek/jaspr), it's a web framework aiming to offer a similar dev experience to Flutter.
If you keep your logic separated, you can even reuse it between your flutter app and a jaspr website.

**Comment by mrgzi** [comment]
Most Flutter web developers complain about not being able to create websites with Flutter. I think this is the last thing flutter should support as a target.

Flutter initially supported mobile devices, but did you advise people to go with electron.js before supporting desktop?

I think instead of discussion to find another framework, it would be worth you to expand your disagreement and so we can think about it if it will worth or not.

**Comment by jsloop42** [comment]
I want to add my view as the proposal was updated. I am looking for using the same code base for web instead of having separate component model for web. That's not want's needed. If that's what it's about then I think Jasper suites well. My requirement is that Flutter web gives option to generate DOM elements instead of using canvas and stuff like that so that I can do UI automation with tools like selenium. I know this is a lot of work. So not sure if it is a valid ask.

For SEO, having ways to improve the metadata, discoverability would help for sites like e-commerce. Have HTML DOM elements will help in this case.

**I am against having a separate component model for web. I want to use the same code base for mobile and web**. Thanks.

**Comment by miyoyo** [comment]
> Most Flutter web developers complain about not being able to create websites with Flutter. I think this is the last thing flutter should support as a target.
> 
> Flutter initially supported mobile devices, but did you advise people to go with electron.js before supporting desktop?

You keep repeating the same point of "flutter doesn't support the web".
It supports the web just as much as it supports other platforms, and you have so far offered zero examples or cases where there is a significant mismatch between platform supports involving the web.

I don't believe this issue is going to go anywhere without these, and I'm not gonna respond without such examples anymore.

**Comment by mrgzi** [comment]
I am not saying flutter doesn't support web, just its not usefull for **websites**. Already gave an example that you can't create such as reddit, nytimes, github etc.. Websites are very essential for companies.

Lets be realistic. Users will never wait engine initialization to see some part of content in websites. This will prevent a lot of users to use Flutter, instead they will go with vuejs and maybe compile cross-platform with tauri. Here is some statistics about user behavior visiting websites; [source](https://www.websitebuilderexpert.com/building-websites/website-load-time-statistics/#:~:text=Pages%20that%20load%20within%20two,Why%20Website%20Speed%20is%20Important)

> - The chance of a bounce increases by 32% when a page load time goes from one to three seconds.
> - On the first page of Google search results, the average page speed is 1.65 seconds.
> - One in four visitors would abandon a website if it takes more than four seconds to load.
> - Pages that load within two seconds have an average bounce rate of 9%, while pages that load in five seconds see their bounce rates skyrocket to 38%.
> - The two-to-three second mark is the turning point where bounce rates skyrocket – 40% of consumers will wait no more than three seconds before abandoning a site.
> - For the top 100 web pages, it only takes an average of 2.5 seconds on desktop and 8.6 seconds on mobile to fully load.
> - 53% of mobile users will abandon a website page if it takes longer than three seconds to load.
> - 82% of consumers say slow page speeds impact their purchasing decisions.
> - 74% of mobile users in the US would leave a mobile site if it didn’t load within five seconds.



Instead of redirecting to another framework, I would rather discuss if this is against of flutter vision or not.

**Comment by iapicca** [comment]
@yjbanov 
why an "offshoot" of a proposal with [>300 likes](https://github.com/flutter/flutter/assets/45637986/55fde978-5b8c-4216-912a-2161044d11e3) in `p3` gets higher priority despite having [more dislike than likes](https://github.com/flutter/flutter/assets/45637986/4e301717-b9d5-403d-80ca-73d129df83e6)?

**Comment by mrgzi** [comment]
@iapicca How do you know this proposal will not get higher likes than dislikes? The issue opened one day ago. 

I also wonder @Hixie 's opinion about this proposal.

**Comment by slavap** [comment]
@mrgzi 
>> Instead of redirecting to another framework, I would rather discuss if this is against of flutter vision or not.

Flutter Web is not designed for "classic" websites - Flutter is about webapps and cross-platform.
It is not possible to "magically" create html/css for Flutter widgets by implementing new renderer.
If you want to develop websites with Dart, consider to use already mentioned https://github.com/schultek/jaspr

**The existing HTML renderer could be improved though** - it is obviously possible to add a more "html friendly" mode (for example "forced" <p> elements for text fields, fewer canvas elements, custom id and class attributes, ,,,). This mode could even be "unusable" for end users, but good enough for server side rendering and SEO.

**Comment by JamesHancockAtteria** [comment]
Here's what I'd like to see (and this may be possible with jaspr/shelf)

1. Integrate shelf for hosting Flutter web.
2. Extend go_router so that it will work with shelf
3. Enable any widget to implement an addon/extension or something, that gives it a new method other than build that is only for rendering simple HTML for bots.
4. Allow basic tags for the bots, and have a hyper-simplistic rendering pipeline that just dumped this as html.  This would still allow for data pull etc. but it would be blocking and could share the same code as the widget itself for pulling data.  The difference is that if shelf detects a bot, then it returns the bot version. If it doesn't, then return flutter web as normal.

Flutter web is fast enough in release mode now that I'm not concerned with load times. I see lots of react ssr stuff that is less responsive.

I am concerned with providing the search engines good content.

**Comment by colbymaloy** [comment]
yeah chiming in on the SEO support as well - there are [Packages](https://pub.dev/packages/seo) that try to help with SEO, but I haven't fully figured out a few issues.

i first built my [Landing Page](https://www.theforestmap.app/) in webflow, and it started ranking on google. theres a button that linked to the [subdomain where my flutter app runs](https://app.theforestmap.app/) and now, thats ranking on google along with help from the meta tags ive set in its index.

The main problem it seems people are having is when it comes to indexing actual CONTENT and the routes which that content is displayed. in my case, if a potential user is on google looking for "pistol location sons of the forest" i want some combination of those words to rank on google, linking to my app page with the content.

Since i've just implemented the SEO packages this week im not sure yet what the crawler is seeing - and quite frankly im so new to SEO that i really dont know what a crawler needs in order to follow a link to go to the next page. I wrapped these images in an Seo.link widget from the package, which inserts an href tag into the html tree. it seems like the package itself updates an html component in the index.html every time it detects one of its 'seo' widgets.
![Screenshot 2024-12-28 at 3 06 31 PM](https://github.com/user-attachments/assets/541ec6e9-1038-4663-b36a-78a1a3ce86c1)

now at this point im assuming the crawler navigates to each of those href links to see what it finds - however, im under the impression that the next flutter route needs to work independently of whether or not user navigates to root domain first.

ex:directly going here doesnt work because the other stuff isnt initialized anywhere but the root domain ie the root '/' : https://app.theforestmap.app/game/sotf
and subsequently 
https://app.theforestmap.app/game/sotf/item/pistol

correct me if im wrong(knowledge limited here) navigator 2.0 was meant to help with this issue and allow direct navigation to nested routes?

but i think for proper indexing, the new page needs to have a its unique <title> tag related to the content on the page. 

the SEO package has a Seo.head widget, but sounds to me like it would only try to add an additional title tag, confusing the bot. 


Either way in theory it sounds like it would work SEO wise to just ensure routes can be navigated to directly, and updating the html metadata tags and having the hrefs for each widget that navigates. am i missing something???? ill come back and update this as i move further along
